model-mesh-ovms-adapter/server/modelmanager.go

```diff
@@ -0,0 +1,504 @@
+					log.Info("Processed UnloadModel for model that was never loaded", "modelId", req.modelId)
+					completeRequest(req, codes.OK, "")
+				} else if modelStatus.State == "END" {
+					completeRequest(req, codes.OK, "")
+				} else {
+					completeRequest(req, codes.Internal, fmt.Sprintf("OVMS model unload failed. state: %s", modelStatus.State))
+				}
+			}
+		}
+
+		// as long as the reload was attempted, assume that the unloadAll completed successfully
+		if unloadAllRequest != nil {
+			completeRequest(unloadAllRequest, codes.OK, "")
+		}
+
+	}
+
+	log.Info("ModelManager thread exiting")
+}
+
+func completeRequest(req *request, code codes.Code, reason string) {
+	// if code == OK, status.Error returns nil
+	req.c <- status.Error(code, reason)
+}
+
+func (mm *OvmsModelManager) getConfig(ctx context.Context) error {
+	// query the Config Status API
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("%s/v1/config", mm.address), http.NoBody)
+	if err != nil {
+		// this should never happen...
+		return fmt.Errorf("Error constructing config status request: %w", err)
+	}
+
+	resp, err := mm.client.Do(req)
+	if err != nil {
+		// TODO: check if error is a timeout and handle appropriately
+		return fmt.Errorf("Protocol error getting the config: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// read the body
+	// NOTE: if the body is not read, the connection cannot be re-used, so
+	// we read the body regardless of the status of the response
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("Error reading config status response body: %w", err)
+	}
+
+	// handle successful request
+	if resp.StatusCode == http.StatusOK {
+		var c OvmsConfigResponse
+		if err1 := json.Unmarshal(body, &c); err1 != nil {
+			return fmt.Errorf("Error parsing config status response: %w", err1)
+		}
+		mm.cachedModelConfigResponse = c
+
+		return nil
+	}
+
+	var errorResponse OvmsConfigErrorResponse
+	if err = json.Unmarshal(body, &errorResponse); err != nil {
+		return fmt.Errorf("Error parsing model config error response: %w", err)
+	}
+
+	errDesc := fmt.Errorf("Error response when getting the config: %s", errorResponse.Error)
+	mm.log.Error(errDesc, "Call to /v1/config returned an error", "code", resp.StatusCode)
+
+	return status.Error(codes.Internal, errDesc.Error())
+}
+
+func (mm *OvmsModelManager) writeConfig() error {
+	// Build the model reposritory config to be written out
+	modelRepositoryConfig := OvmsMultiModelRepositoryConfig{
+		ModelConfigList: make([]OvmsMultiModelConfigListEntry, len(mm.loadedModelsMap)),
+	}
+	listIndex := 0
+	for _, model := range mm.loadedModelsMap {
+		modelRepositoryConfig.ModelConfigList[listIndex] = model
+		listIndex++
+	}
+
+	modelRepositoryConfigJSON, err := json.Marshal(modelRepositoryConfig)
+	if err != nil {
+		return fmt.Errorf("Error marshalling config file: %w", err)
+	}
+
+	if err := ioutil.WriteFile(mm.modelConfigFilename, modelRepositoryConfigJSON, 0644); err != nil {
```

Comment:
I don't think an atomic swap would be needed. At least in the ServingRuntime I've been testing with I pass `--file_system_poll_wait_seconds=0` to the OVMS container so that it isn't doing its own polling. Tracking revision history may be useful though (and then atomic swaps comes with it).

I too am unsure of how to handle a total failure in the config reloading like this (i.e. a case where `updateModelConfig` returns an error), it leaves everything in an uncertain state and it would be hard to determine if it could recover on its own.